lib/sorbet-rails/model_plugins/active_record_assoc.rb

```diff
@@ -56,13 +56,25 @@ def populate_single_assoc_getter_setter(assoc_module_rbi, assoc_name, reflection
     # optional (via `optional` or `!required` or `!belongs_to_required_by_default`)
     return false if !reflection.belongs_to?
 
-    if reflection.options.key?(:required)
-      !!reflection.options[:required]
-    elsif reflection.options.key?(:optional)
-      !reflection.options[:optional]
-    else
-      !!reflection.active_record.belongs_to_required_by_default
+    rails_required_config =
+      if reflection.options.key?(:required)
+        !!reflection.options[:required]
+      elsif reflection.options.key?(:optional)
+        !reflection.options[:optional]
+      else
+        !!reflection.active_record.belongs_to_required_by_default
+      end
+
+    column_def = @columns_hash[reflection.foreign_key.to_s]
+    db_required_config = column_def && !column_def.null
+
+    if rails_required_config && !db_required_config
+      puts "Warning: belongs_to association #{reflection.name} is required at the application
+        level but not at the DB level. It's recommended to enforce it by adding null & foreign key
+        constraint at the DB level.".squish!
     end
```

Comment:
how do you feel about also adding a warning for the inverse case?

```ruby
    elsif !rails_required_config && db_required_config
      puts "Warning: belongs_to association #{reflection.name} is required at the DB
        level but not at the application level. Add a constraint at the app level (using `optional: false`)
        as a validation hint to Rails.".squish!
    end
```

Reply:
I can, but when you do both, it also means doing extra unnecessary validation. What do you think?
 
I consider this direction of the warning because the DB level is the most certain level to put any kind of restriction (application level is not sufficient). Having both probably means doing extra work.

Reply:
Looking at the implementation of `belongs_to` and `validates_presence_of`, I think it's worthy to have both: the validation is simple (`.blank?`) and probably cost a little to do. Lack of it means people may not be truly aware of the required-ness of the association. 
I'll add a warning whenever the two configurations don't match.

Reply:
We've been removing `validates_presence_of` in favour of `belongs_to optional: false` because of the sorbet-rails benefits it gives. Thank you for adding this!